File: app/src/main/java/com/vladislav/workoutassistant/ui/workouts/viewmodels/WorkoutViewModel.java
package com.vladislav.workoutassistant.ui.workouts.viewmodels;

import android.app.Application;

import com.vladislav.workoutassistant.data.DataRepository;
import com.vladislav.workoutassistant.data.db.entity.Exercise;
import com.vladislav.workoutassistant.data.models.Item;
import com.vladislav.workoutassistant.data.models.WorkoutExercise;
import com.vladislav.workoutassistant.data.models.WorkoutProgram;
import com.vladislav.workoutassistant.data.models.WorkoutSet;

import java.util.ArrayList;
import java.util.List;

import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MediatorLiveData;
import androidx.lifecycle.Observer;

public class WorkoutViewModel extends AndroidViewModel {

    private DataRepository mDataRepository;
    private WorkoutProgram mWorkoutProgram;
    private List<Item> mMuscleGroups;
    private MediatorLiveData<List<Object>> mSetsAndExercises = new MediatorLiveData<>();

    public WorkoutViewModel(Application application) {
        super(application);
        mDataRepository = DataRepository.Companion.getInstance(application);
        mMuscleGroups = mDataRepository.loadMuscleGroups();
    }

    public void init(final int workoutId) {
        mSetsAndExercises.addSource(mDataRepository.loadWorkoutProgram(workoutId), new Observer<WorkoutProgram>() {
            @Override
            public void onChanged(final WorkoutProgram program) {
                final List<WorkoutSet> workoutSets = program.getWorkoutSets();
                if (workoutSets != null) {
                    ArrayList<Integer> exerciseIds = new ArrayList<>();
                    for (WorkoutSet set : workoutSets) {
                        for (WorkoutExercise workoutExercise : set.getWorkoutExercises()) {
                            exerciseIds.add(workoutExercise.matchingInfo.getExerciseId());
                        }
                    }
                    mSetsAndExercises.addSource(mDataRepository.loadExercises(exerciseIds), new Observer<List<Exercise>>() {   //TODO: how to work in background?
                        @Override
                        public void onChanged(List<Exercise> exercises) {
                            if (exercises != null) {
                                for (WorkoutSet workoutSet : workoutSets) {
                                    addExerciseDataToSet(workoutSet.getWorkoutExercises(), exercises);
                                }
                                mSetsAndExercises.postValue(workoutProgramToList(program));
                                mWorkoutProgram = program;
                            }
                        }
                    });
                }
            }
        });
    }

    public List<Item> getMuscleGroups() {
        return mMuscleGroups;
    }

    public WorkoutProgram getWorkoutProgram() {
        return mWorkoutProgram;
    }

    public LiveData<List<Object>> getSetsAndExercisesList() {
        return mSetsAndExercises;
    }

    private List<Object> workoutProgramToList(WorkoutProgram program) {
        List<Object> list = new ArrayList<>();
        List<WorkoutSet> sets = program.getWorkoutSets();
        for (WorkoutSet set : sets) {
            list.add(set);
            list.addAll(set.getWorkoutExercises());
        }
        return list;
    }

    private void addExerciseDataToSet(List<WorkoutExercise> workoutExercises, List<Exercise> exercises) {
        for (WorkoutExercise workoutExercise : workoutExercises) {
            int keyId = workoutExercise.matchingInfo.getExerciseId();
            for (Exercise exercise : exercises) {
                if (exercise.getId() == keyId) {
                    workoutExercise.exerciseInfo = exercise;
                    break;
                }
            }
        }
    }
}

File: app/src/main/java/com/vladislav/workoutassistant/ui/workouts/viewmodels/WorkoutViewModel.kt
package com.vladislav.workoutassistant.ui.workouts.viewmodels

import android.app.Application
import androidx.lifecycle.*

import com.vladislav.workoutassistant.data.DataRepository
import com.vladislav.workoutassistant.data.db.entity.Exercise
import com.vladislav.workoutassistant.data.models.*

import java.util.ArrayList

class WorkoutViewModel(application: Application) : AndroidViewModel(application) {

    private val mDataRepository: DataRepository = DataRepository.getInstance(application)
    private val mWorkoutInfo = MediatorLiveData<WorkoutInfo>()

    val muscleGroups: List<Item> = mDataRepository.loadMuscleGroups()
    val workoutInfo: LiveData<WorkoutInfo>
        get() = mWorkoutInfo

    var workoutProgram: WorkoutProgram? = null
        private set

    private val mSetsAndExercises = MediatorLiveData<List<Any>>()

    val setsAndExercisesList: LiveData<List<Any>>
        get() = mSetsAndExercises

    fun init(workoutId: Int) {
        mSetsAndExercises.addSource(mDataRepository.loadWorkoutProgram(workoutId)) { program ->
            val workoutSets = program.workoutSets
            if (workoutSets != null) {
                val exerciseIds = ArrayList<Int>()
                for (set in workoutSets) {
                    for (workoutExercise in set.workoutExercises!!) {
                        exerciseIds.add(workoutExercise.matchingInfo!!.exerciseId)
                    }
                }
                mSetsAndExercises.addSource(mDataRepository.loadExercises(exerciseIds)) { exercises ->
                    //TODO: how to work in background?
                    if (exercises != null) {
                        for (workoutSet in workoutSets) {
                            addExerciseDataToSet(workoutSet.workoutExercises!!, exercises)
                        }
                        mSetsAndExercises.postValue(workoutProgramToList(program))
                        workoutProgram = program
                    }
                }
            }
        }
    }

    private fun workoutProgramToList(program: WorkoutProgram): List<Any> {
        val list = ArrayList<Any>()
        val sets = program.workoutSets
        for (set in sets!!) {
            list.add(set)
            list.addAll(set.workoutExercises!!)
        }
        return list
    }

    private fun addExerciseDataToSet(workoutExercises: List<WorkoutExercise>, exercises: List<Exercise>?) {
        for (workoutExercise in workoutExercises) {
            val keyId = workoutExercise.matchingInfo!!.exerciseId
            for (exercise in exercises!!) {
                if (exercise.id == keyId) {
                    workoutExercise.exerciseInfo = exercise
                    break
                }
            }
        }
    }

    fun loadWorkoutInfoById(id: Int) {

        fun extractExercisesIds(sets: List<SetInfo>): List<Int> {
            val ids = ArrayList<Int>()
            for (set in sets) {
                for (exercise in set.exercises) {
                    ids.add(exercise.id)
                }
            }
            return ids
        }

        fun addExerciseContent(sets: List<SetInfo>, content: List<ExerciseContent>) {
            for (set in sets) {
                for (exercise in set.exercises) {
                    val exerciseContent = content.find { it.id == exercise.id }
                    exercise.name = exerciseContent?.name ?: ""
                    exercise.muscleGroupId = exerciseContent?.muscleGroupId ?: 0
                }
            }
        }

        mWorkoutInfo.addSource(mDataRepository.loadWorkoutInfo(id)) { info ->
            val ids: List<Int> = extractExercisesIds(info.sets)
            mWorkoutInfo.addSource(mDataRepository.loadExercisesById(ids)) { content ->
                addExerciseContent(info.sets, content)
                mWorkoutInfo.postValue(info)
            }
        }
    }
}
